File: lab_assignments/lab8/heap-sort.c
#include <stdio.h>
#include <stdlib.h>

void heapify(int arr[], int n, int i)
{
    int largest = i;  // Initialize largest as root
    int l = 2*i + 1;  // left = 2*i + 1
    int r = 2*i + 2;  // right = 2*i + 2
 
    // If left child is larger than root
    if (l < n && arr[l] > arr[largest])
        largest = l;
 
    // If right child is larger than largest so far
    if (r < n && arr[r] > arr[largest])
        largest = r;
 
    // If largest is not root
    if (largest != i)
    {
        swap(arr[i], arr[largest]);
 
        // Recursively heapify the affected sub-tree
        heapify(arr, n, largest);
    }
}
 
// main function to do heap sort
void heapSort(int arr[], int n)
{
	int i;
    // Build heap (rearrange array)
    for ( i = n / 2 - 1; i >= 0; i--)
        heapify(arr, n, i);
 
    // One by one extract an element from heap
    for (i=n-1; i>=0; i--)
    {
        // Move current root to end
        swap(arr[0], arr[i]);
 
        // call max heapify on the reduced heap
        heapify(arr, i, 0);
    }
}

int main() {

int n,i;
	scanf("%d",&n);
	int arr[n];

	for(i = 0; i < n; i++)
		scanf("%d",&arr[i]);

	heap_sort(arr, n);

	for(i = 0; i < n; i ++)
		printf("%d ",arr[i]);

	return 0;
}

File: Algorithms/Sort_Algorithms/Heap_Sort/heapsort.py
def maxHeapify(a,i,size):
	large = i
	left = 2*i+1
	right = 2*i+2
	if(left < size and a[left]>a[i]):
		large = left
	if(right < size and a[right]>a[large]):
		large = right
	if(large != i):
		a[i],a[large]=a[large],a[i]
		maxHeapify(a,large,size)
def buildMaxHeap(a,size):
	for i in range((size//2)-1,-1,-1):
		maxHeapify(a,i,size)

def heapSort(a):
	size=len(a)
	buildMaxHeap(a,size)
	for i in range(size-1,0,-1):
		a[i],a[0]=a[0],a[i]		# Swap Array[0] with Array[i]
		maxHeapify(a,0,i)
a = [4,2,1,6,3,6,7,9,-8,0]
heapSort(a)
print(a)
